clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -12408,6 +12408,29 @@ def err_sycl_special_type_num_init_method : Error<
   "types with 'sycl_special_class' attribute must have one and only one '__init' "
   "method defined">;
 
+// SYCL kernel entry point diagnostics
+def err_sycl_entry_point_invalid : Error<
+  "'sycl_kernel_entry_point' attribute cannot be applied to a"
+  " %select{non-static member|variadic|deleted|defaulted|constexpr|consteval|"
+           "noreturn|coroutine}0 function">;
+def err_sycl_entry_point_invalid_redeclaration : Error<
+  "'sycl_kernel_entry_point' kernel name argument does not match prior"
+  " declaration%diff{: $ vs $|}0,1">;
+def err_sycl_kernel_name_conflict : Error<
+  "'sycl_kernel_entry_point' kernel name argument conflicts with a previous"
+  " declaration">;
+def warn_sycl_kernel_name_not_a_class_type : Warning<
+  "%0 is not a valid SYCL kernel name type; a class type is required">,
```

Comment:
Do we mean 'class type'? As that includes `union`.  Also, I find myself wondering, do the restrictions make sense to exclude scoped-enum?

Reply:
That would be acceptable, I might consider `class or struct type`, which doesn't get into "what C++ has to say about it" by not using a term of art, and seems similarly descriptive.

Reply:
Done. I stuck with "non-union class" since the SYCL specification just says "class". We can revisit if/when the SYCL specification is updated to better specify its intent.